add functionality for different calnex firmware file formats

PDeanCalnex opened the pull request:
## Summary

Currently the code only works for a specific string of the filename. 
This can be different depending on the release that is sent. If we are doing an engineering release or full release. 
We normally now send files of the format calnex_combined_fw now that we have installer files for both sentinel and sentry included in the file

## Test Plan
[dev@pd-vm01.sen.local] [~/meta-time] (main *$%<>)
$ ./cmd/calnex/calnex firmware --device sen050.sen.local --file sentinel_fw_v2.21.2.0.tar --insecureTLS
WARN[0000] sen050.sen.local: is running 21.0.0.9718-d-20241125, latest is 21.2.0. Needs an update 
INFO[0000] sen050.sen.local: dry run. Not upgrading firmware 

[dev@pd-vm01.sen.local] [~/meta-time] (main *$%<>)
$ ./cmd/calnex/calnex firmware --device sen050.sen.local --file calnex_combined_fw_R21.2.0.tar --insecureTLS
WARN[0000] sen050.sen.local: is running 21.0.0.9718-d-20241125, latest is 21.2.0. Needs an update 
INFO[0000] sen050.sen.local: dry run. Not upgrading firmware 


successfully extracting the correct part of the firmware from each differently formatted file name

leoleovich commented on calnex/firmware/ossfirmware.go:
```diff
@@ -31,12 +32,34 @@ type OSSFW struct {
 	basename := filepath.Base(fw.filepath)
-	vs := strings.ReplaceAll(strings.TrimSuffix(basename, filepath.Ext(basename)), "sentinel_fw_v", "")
-	v, err := version.NewVersion(strings.ToLower(vs))
+	// Extract version from filename
+	// sentinel_fw_v2.13.1.0.5583D-20210924.tar
+	// calnex_combined_fw_R21.0.0.9705-20241111.tar
+	if strings.HasPrefix(basename, "sentinel_fw_") {
+		vs = strings.ReplaceAll(strings.TrimSuffix(basename, filepath.Ext(basename)), "sentinel_fw_", "")
+	} else if strings.HasPrefix(basename, "sentry_fw_") {
+		vs = strings.ReplaceAll(strings.TrimSuffix(basename, filepath.Ext(basename)), "sentry_fw_", "")
+	} else if strings.HasPrefix(basename, "calnex_combined_fw_") {
+		vs = strings.ReplaceAll(strings.TrimSuffix(basename, filepath.Ext(basename)), "calnex_combined_fw_", "")
+	} else {
+		return nil, fmt.Errorf("Unexpected file string, expected sentinel_fw_ sentry_fw_ or calnex_combined_fw_  at start: %s", basename)
+	}
+
+	if strings.HasPrefix(vs, "v") {
+		v, err = version.NewVersion(strings.SplitN(strings.ToLower(vs), ".", 2)[1])
+	} else if strings.HasPrefix(vs, "R") {
+		v, err = version.NewVersion(strings.TrimPrefix(strings.ToLower(vs), "r"))
+	} else {
+		return nil, fmt.Errorf("Unexpected file string, expected v2. or R. : %s", vs)
+	}
```
Can you please cover this beautiful logic with unit tests so we never break it by mistake?

PDeanCalnex commented:
That should be unit tests covering all the file formats we use and checking we always get the right output back

deathowl commented on calnex/firmware/ossfirmware.go:
```diff
@@ -31,12 +32,34 @@ type OSSFW struct {
+	} else if strings.HasPrefix(basename, "sentry_fw_") {
+		vs = strings.ReplaceAll(strings.TrimSuffix(basename, filepath.Ext(basename)), "sentry_fw_", "")
+	} else if strings.HasPrefix(basename, "calnex_combined_fw_") {
+		vs = strings.ReplaceAll(strings.TrimSuffix(basename, filepath.Ext(basename)), "calnex_combined_fw_", "")
```
you could put these prefixes into a regex with or like https://go.dev/play/p/oiqttenXhFO

PDeanCalnex replied:
yes much neater. Implemented

deathowl commented on calnex/firmware/ossfirmware.go:
```diff
@@ -31,12 +33,33 @@ type OSSFW struct {
+		return nil, fmt.Errorf("invalid filename, expected prefix sentinel_fw_, sentry_fw_, or calnex_combined_fw_: %s", basename)
+	}
+
+	if strings.HasPrefix(vs, "v") {
```
can  we somehow make this part of the regex solution above. It would still eliminate a lot of if-else-ing

PDeanCalnex replied:
I had thought it would just make it messy but yes I think the following works. 

    re, _ := regexp.Compile("(sentry_fw_|sentinel_fw_|calnex_combined_fw_)(v[0-9].|R)(.*).tar")
    
I don't have time to test it out properly today though.

PDeanCalnex commented:
Fixed up RegEx to be more complete. Removed some unnecessary definitions at the start of the function. 

I've been rebasing all the commits each time. Do you actually want me to rebase each time I push any changes? Or just to before I open a PR and then when it's ready to merge?

leoleovich commented on calnex/firmware/ossfirmware.go:
```diff
@@ -31,11 +33,25 @@ type OSSFW struct {
+
+	re := regexp.MustCompile(`(sentry_fw_|sentinel_fw_|calnex_combined_fw_)(v[0-9]\.|R)(.*).tar`)
+
+	matches := re.FindStringSubmatch(basename)
+	if matches != nil {
+		vs = matches[3]
```
Just to be safe I would add a check for `len(matches) == 4` to avoid panic. And while at it add a test case scenario with a string which doesn't match the regexp (check the error handling)

leoleovich commented:
Last comment - otherwise looks good! Thank you for your contribution!

leoleovich commented:
Also please make sure you have only 1 commit (squash them)

deathowl commented:
LGTM, but please fix lint
  Error: File is not `gofmt`-ed with `-s` (gofmt)

PDeanCalnex commented:
> LGTM, but please fix lint Error: File is not `gofmt`-ed with `-s` (gofmt)

I think I now have the same linter as you working in my VS code so hopefully no more lint issues. 
Our standard VMs version of ubuntu has an old default go install so that was causing me some issues
